packages/xstate-wallet/src/workflows/close-ledger-and-withdraw.ts

```diff
@@ -1,87 +1,173 @@
-};
-type WorkflowGuards = {
-  doesChannelIdExist: ConditionPredicate<WorkflowContext, WorkflowEvent>;
-};
-interface WorkflowActions extends ActionFunctionMap<WorkflowContext, WorkflowEvent> {
-  sendResponse: ActionFunction<WorkflowContext, WorkflowEvent>;
-  assignLedgerId: ActionFunction<WorkflowContext, DoneInvokeEvent<CloseLedger>>;
-  hideUi: ActionFunction<WorkflowContext, any>;
-  displayUi: ActionFunction<WorkflowContext, any>;
-  clearBudget: ActionFunction<WorkflowContext, any>;
+enum Actions {
+  sendResponse = 'sendResponse',
+  assignLedgerId = 'assignLedgerId',
+  setTransactionId = 'setTransactionId',
+  assignBudget = 'assignBudget'
 }
-interface WorkflowServices extends Record<string, ServiceConfig<WorkflowContext>> {
-  getFinalState: (context: WorkflowContext, event: WorkflowEvent) => Promise<SupportState.Init>;
-  supportState: StateMachine<any, any, any>;
-  submitWithdrawTransaction: (context: WorkflowContext, event: WorkflowEvent) => Promise<void>;
-  createObjective: (context: WorkflowContext, event: any) => Promise<Objective>;
+
+enum Services {
+  constructFinalState = 'constructFinalState',
+  supportState = 'supportState',
+  submitWithdrawTransaction = 'submitWithdrawTransaction',
+  createObjective = 'createObjective',
+  observeFundsWithdrawal = 'observeFundsWithdrawal',
+  fetchBudget = 'fetchBudget',
+  clearBudget = 'clearBudget'
 }
 
-export const config: StateNodeConfig<WorkflowContext, any, any> = {
+export type WorkflowActions = CommonWorkflowActions &
+  Record<
+    Actions,
+    ActionFunction<WorkflowContext, WorkflowEvent> | AssignAction<WorkflowContext, WorkflowEvent>
+  >;
+export type WorkflowServices = Record<Services, ServiceConfig<WorkflowContext>>;
+
+export const config: MachineConfig<WorkflowContext, any, WorkflowEvent> = {
   id: 'close-and-withdraw',
-  initial: 'waitForUserApproval',
+
+  initial: 'fetchBudget',
   states: {
+    fetchBudget: {
+      invoke: {
+        src: Services.fetchBudget,
+        onDone: {target: 'waitForUserApproval', actions: [Actions.assignBudget]}
+      }
+    },
     waitForUserApproval: {
-      entry: ['displayUi'],
-
+      entry: [CommonActions.displayUI],
       on: {
-        USER_APPROVES_CLOSE: [
-          {cond: 'doesChannelIdExist', target: 'closeLedger'},
-          'createObjective'
-        ],
-        USER_REJECTS_CLOSE: {target: 'failure'}
+        USER_APPROVES_CLOSE: {target: 'createObjective'},
+        USER_REJECTS_CLOSE: {target: 'userDeclinedFailure'}
       }
     },
 
     createObjective: {
       invoke: {
-        src: 'createObjective',
-        onDone: {target: 'closeLedger', actions: ['assignLedgerId']}
+        src: Services.createObjective,
+        onDone: {target: 'closeLedger', actions: [Actions.assignLedgerId]}
       }
     },
-    closeLedger: getDataAndInvoke({src: 'getFinalState'}, {src: 'supportState'}, 'withdraw'),
+    closeLedger: getDataAndInvoke<LedgerExists>(
+      {src: Services.constructFinalState},
```

Comment:
Would be an option here to separate these out into two steps in this machine, and store the final state on this machine itself. It seems like that would give you slightly better control if you wanted to handle the case where one of these steps failed, in terms of what you show to the user / the next actions you take.